src/shogun2-core/shogun2-service/src/main/java/de/terrestris/shogun2/service/PasswordResetTokenService.java

```diff
@@ -0,0 +1,370 @@
+				Restrictions.eq("token", token)
+		);
+
+		PasswordResetToken passwordResetToken =
+				dao.findByUniqueCriteria(criteria);
+
+		return passwordResetToken;
+	}
+
+	/**
+	 *
+	 * @param request
+	 * @param email
+	 * @return
+	 * @throws UsernameNotFoundException
+	 * @throws Exception
+	 * @throws URISyntaxException
+	 */
+	public Boolean sendResetPasswordMail(HttpServletRequest request, String email) throws
+			UsernameNotFoundException, Exception, URISyntaxException {
+
+		Boolean success = false;
+
+		// get the user by the provided email address
+		User user = userService.findByEmail(email);
+
+		if (user == null) {
+			throw new UsernameNotFoundException("Could not find user "
+					+ "with the provided email: " + email);
+		}
+
+		// generate and save the unique reset-password token for the user
+		PasswordResetToken resetPasswordToken = generateResetPasswordToken(user);
+
+		// create the reset-password URI that will be send to the user
+		URI resetPasswordURI = createResetPasswordURI(request,
+				resetPasswordToken);
+
+		// create a thread safe "copy" of the template message
+		SimpleMailMessage resetPwdMsg = new SimpleMailMessage(
+				resetPasswordMailMessageTemplate);
+
+		// prepare a personalized mail with the given token
+		resetPwdMsg.setTo(email);
+		resetPwdMsg.setText(
+				String.format(
+						resetPwdMsg.getText(),
+						user.getFirstName(),
+						user.getLastName(),
+						UriUtils.decode(resetPasswordURI.toString(), "UTF-8")
+				)
+		);
+
+		// and send the mail
+		mailPublisher.sendMail(resetPwdMsg);
+
+		// we did it!
+		success = true;
+
+		return success;
+	}
+
+	/**
+	 *
+	 * @param password
+	 * @param id
+	 * @param token
+	 * @return
+	 * @throws Exception
+	 */
+	public Boolean changePassword(String password, int id, String token)
+			throws Exception {
+
+		Boolean success = false;
+
+		// try to find the provided token
+		PasswordResetToken passwordResetToken = findByIdAndToken(id, token);
+
+		if (passwordResetToken == null) {
+			throw new Exception("The provided token is not valid.");
+		}
+
+		DateTime expirationDate = (DateTime) passwordResetToken
+				.getExpirationDate();
+
+		// check if the token expire date is valid
+		if (expirationDate.isBeforeNow()) {
+			throw new Exception("The provided token is expired.");
+		}
+
+		// the provided token seems to be valid, the user's password can be
+		// be changed
+
+		// get the user by the provided token
+		User user = passwordResetToken.getUser();
+
+		if (user == null) {
+			throw new Exception("Could not find the user for the "
+					+ "provided token.");
+		}
+
+		// finally update the password (encrypted)
+		try {
+			user.setPassword(passwordEncoder.encode(password));
+			userService.updateExistingUser(user);
+			LOG.debug("Successfully updated the password.");
+		} catch(Exception e) {
+			throw new Exception("Could not update the password: "
+					+ e.getMessage());
+		}
+
+		// delete the token
+		dao.delete(passwordResetToken);
```

Comment:
Should this be moved to the above try-catch in a finally? When there was a problem with the token, we simply throw and the token will never be removed, right?